Add draft Multichain API MIPs (#54)
MetaMask/metamask-improvement-proposals: MIPs/mip-x-multichainapi.md

---
mip: x
title: Adopt chain agnostic standards for a Multichain API
status: Draft
stability: n/a
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/53
author(s): Alex Donesky (@adonesky1), Jiexi Luan (@jiexi), Vandan Parikh(@vandan) 
type: Maintainer
created: 2024-10-28
---

## Summary

> This proposal recommends that MetaMask adopt a set of Chain Agnostic Improvement Proposals (CAIPs) to guide the design of a new Wallet API that is generalized for interactions between dapps and wallets in a multichain environment. A "Multichain API". Implementing these standards encourages broad adoption while enhancing interoperability, improving user experiences, and significantly streamlining development for multichain dapps.

Reviewer: ```suggestion
This proposal recommends that MetaMask adopt a set of Chain Agnostic Improvement Proposals ([CAIPs](https://github.com/ChainAgnostic/CAIPs)) to guide the design of a new Wallet API that is generalized for interactions between dapps and wallets in a multichain environment. A "Multichain API". Implementing these standards encourages broad adoption while enhancing interoperability, improving user experiences, and significantly streamlining development for multichain dapps.
```

### Goals
- Consistent Structure: Provide common patterns while separating namespaces of each ecosystem or network.
- Interoperability: Employ standards to encourage the development of powerful libraries with support for multiple wallets and networks
- Unlock the Design Space: Enable dapps to negotiate an interface with wallets for optimal multichain experiences.

> - Security: Enhance security and privacy associated with wallets delived through Browser Extensions.

Reviewer: Kind of mentioned by the above points, but thoughts on `Extensibility` as it's own item? nbd

Author: Yeah, I think we should highlight the goal of improved Extensibility here. I'll adjust the bullet points 👍

## Motivation
Valuable lessons have been learned from scaling the Ethereum dapp and wallet ecosystem. Now, a subset of [CAIPs](https://github.com/ChainAgnostic/CAIPs) offer a path to dramatically improve upon the prevailing patterns that had been established through [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) and [EIP-3326](https://eips.ethereum.org/EIPS/eip-3326) with chain-agnostic patterns that address long-standing issues. Furthermore, as decentralized networks continue to diversify, these patterns can be applied to non-EVM ecosystems so they can scale more efficiently and avoid common pain points associated with the demands of a growing dapp ecosystem that may require rapid evolution of Wallet APIs.

The Multichain API and associated CAIP standards promise to enable the following benefits for developers:

> - Ability to interact with non-EVM decentralized networks through a consistent interface

Reviewer: ```suggestion
- Ability to interact with EVM and non-EVM decentralized networks through a consistent interface
```

worth adding this? or is it already obviously implied?

> **Note:** A `sessionId` will not be included as part of the event.

### CAIP-285 - Revoke Authorization Scopes
A dapp can make a [CAIP-285](https://chainagnostic.org/CAIPs/caip-285) `wallet_revokeSession` request to revoke all of its authorization scopes.

> **Note:** a `sessionId` parameter is not supported and will be ignored, if included.

### API Specification Document
An OpenRPC [specification for the Multichain API](https://github.com/MetaMask/api-specs/blob/main/multichain/openrpc.yaml) can be found in the [api-specs](https://github.com/MetaMask/api-specs) repository.

## Caveats
The CAIPs referenced in this proposal are still in Draft or Review status with CASA and may be subject to material changes.

### API Access
As part of the Multichain API, MetaMask is leading the adoption of alternative methods for browser extension wallet communication. The current practice for browser extension wallets is to inject a javascript object in order to communicate with each site. But this approach comes with privacy, security, reliability, and performance drawbacks.

> - The MetaMask Browser Extension will adopt an alternative method of communication through the `externally_connectable` feature, which is now supported by most modern web browsers with extension frameworks.

Reviewer: ```suggestion
- The MetaMask Browser Extension will adopt an alternative method of communication through the [externally_connectable](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable) feature, which is now supported by most modern web browsers with extension frameworks.
```

> - When a site is accessed through MetaMask's Mobile In-app Browser, the Multichain API will be available through an injected `window.wallet` object. 

Reviewer: guessing this is placeholder, as in we haven't actually decided on the details of this yet

Author: yes, placeholder for now.

Author: Todo: Add a note about Firefox & Mobile.

In summary, the following methods will solely be accessible through the new communication methods:
`wallet_createSession` (CAIP-25)
`wallet_sessionChanged` (CAIP-311)
`wallet_getSession` (CAIP-312)
`wallet_revokeSession` (CAIP-285)
`wallet_invokeMethod` (CAIP-27)
`wallet_notify` (CAIP-319)

(API Delivery & Wallet Discovery CAIPs to be added in this section)

### Snap Methods
Some RPC APIs associated with Snaps will be rendered redundant and will not be supported through the Multichain API. However, they will remain accessible through the EIP-1193 interface for backward compatibility.

**Discontinued Methods:**
wallet_requestSnaps
wallet_getSnaps
wallet_invokeSnap
wallet_snap

## Implementation
API Maintainers will implement the multichain interface in coordination with multiple MetaMask teams.
The Multichain API is intended to interoperate with MetaMask Snaps. See [SIP-26](https://github.com/MetaMask/SIPs/blob/ed17dd33713e6c2203f11b85ba655ae4acbcca7a/SIPS/sip-26.md) for a high-level overview of the anticipated architectural approach.

## Developer Adoption Considerations

> dapp developers are encouraged to use libraries to abstract the connection differences between wallet clients and to ensure their dapp stays updated as wallet-to-dapp communication methods continue to evolve.

Reviewer: ```suggestion
Dapp developers are encouraged to use libraries to abstract the connection differences between wallet clients and to ensure their dapp stays updated as wallet-to-dapp communication methods continue to evolve.
```

---
mip: x
title: Adopt chain agnostic standards for a Multichain API
status: Draft
stability: n/a
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/53
author(s): Alex Donesky (@adonesky1), Jiexi Luan (@jiexi), Vandan Parikh(@vandan) 
type: Maintainer
created: 2024-10-28
---

## Summary

> This proposal recommends that MetaMask adopt a set of Chain Agnostic Improvement Proposals ([CAIPs](https://github.com/ChainAgnostic/CAIPs)) to guide the design of a new Wallet API that is generalized for interactions between dapps and wallets in a multichain environment. A "Multichain API". Implementing these standards encourages broad adoption while enhancing interoperability, improving user experiences, and significantly streamlining development for multichain dapps.

Reviewer 2: ```suggestion
This proposal recommends that MetaMask adopt a set of Chain Agnostic Improvement Proposals ([CAIPs](https://github.com/ChainAgnostic/CAIPs)) to guide the design of a new Wallet API that is generalized for interactions between dapps and wallets in a multichain environment. Implementing these standards encourages broad adoption while enhancing interoperability, improving user experiences, and significantly streamlining development for multichain dapps.
```

Author: Do you think it's best to leave that in the Definitions section?

---
mip: x
title: Adopt chain agnostic standards for a Multichain API
status: Draft
stability: n/a
discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions/53
author(s): Alex Donesky (@adonesky1), Jiexi Luan (@jiexi), Vandan Parikh(@vandan) 
type: Maintainer
created: 2024-10-28
---

## Summary
This proposal recommends that MetaMask adopt a set of Chain Agnostic Improvement Proposals ([CAIPs](https://github.com/ChainAgnostic/CAIPs)) to guide the design of a new Wallet API that is generalized for interactions between dapps and wallets in a multichain environment. Implementing these standards encourages broad adoption while enhancing interoperability, improving user experiences, and significantly streamlining development for multichain dapps.

### Goals
- Consistent Structure: Provide common patterns while separating namespaces of each ecosystem or network.
- Interoperability: Employ standards to encourage the development of powerful libraries with support for multiple wallets and networks.
- Unlock the Design Space: Enable dapps to negotiate an interface with wallets for optimal multichain experiences.

> - Extensibility: Make the API reasonbaly future-proof, allowing for new networks and methods to be added as demanded by the ecosystem.
> - Security: Enhance security and privacy associated with wallets delived through Browser Extensions.

Reviewer 3: ```suggestion
- Extensibility: Make the API reasonably future-proof, allowing for new networks and methods to be added as demanded by the ecosystem.
- Security: Enhance security and privacy associated with wallets delivered through Browser Extensions.
```

The Multichain API and associated CAIP standards promise to enable the following benefits for developers:
- Ability to interact with EVM and non-EVM decentralized networks through a consistent interface
- Employ interface negotiation patterns to adopt novel wallet features, while gracefully degrading for wallets that may not yet support them
- Ability to simultaneously interact with multiple networks
- Be notified of updates across multiple networks
- Elimination of excessive error handling involved with chain-switching across EVM networks

This proposal also generally encourages a privacy preserving and more secure alternative to provider injection (ex. `window.<provider>`) to establish communication between sites and browser extension wallets. 

## Usage Example

### Request
An example structure for a JSON-RPC request that a dapp would send to request authorization(s) from MetaMask:

```js
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_createSession",
  "params": {
    "optionalScopes": {
      "wallet:eip155": {

>         "methods": ["wallet_addEthereumChain"],

Reviewer: ```suggestion
        "methods": ["wallet_addEthereumChain"],
        "notifications": [],
```

      },
      "eip155:1": {
        "methods": ["eth_sendTransaction","eth_call","eth_getBalance","eth_blockNumber","eth_getTransactionCount","wallet_watchAsset","eth_subscribe"],
        "notifications": ["eth_subscription"],
      },
      "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp": {
        "methods": ["signAndSendTransaction","signAllTransactions","signMessage"],

>         "notifications": ["accountChanged"],

Reviewer: leave as empty array maybe?

      }
      "bip122:000000000019d6689c085ae165831e93": {
        ...
      }
    },
    "scopedProperties": {
      // Scope-keyed proposals or declarations
      "eip155:1": {
        "extension_foo": "bar"    
      }
    },
  },
}
```

### Response
An example structure for the corresponding JSON-RPC response that a dapp would receive from the wallet:

```js
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "sessionScopes": { 
      "eip155:wallet": {

>           "methods": ["wallet_addEthereumChain"],

Reviewer: ```suggestion
          "methods": ["wallet_addEthereumChain"],
          "notifications": [],
```

notifications is required

          "accounts": ["eip155:wallet:0x0910e12C68d02B561a34569E1367c9AAb42bd810"]
      },
      "eip155:1": {
        "methods": ["eth_sendTransaction","eth_call","eth_getBalance","eth_blockNumber","eth_getTransactionCount","wallet_watchAsset","eth_subscribe"],
        "notifications": ["eth_subscription"],
        "accounts": ["eip155:1:0x0910e12C68d02B561a34569E1367c9AAb42bd810"]
      },
      "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp": {
        "methods": ["signAndSendTransaction","signAllTransactions","signMessage"],

>         "notifications": ["accountChanged"],

Reviewer: ```suggestion
        "notifications": [],
```

maybe we just leave this blank in case this mock notification name confuses someone from the solona space

Reviewer: idk. nbd

Author: afaik, this notification example is one that some wallets, such as Phantom, do implement.

Author: On second thought. I'll remove to avoid potential confusion in case this notification isn't broadly adopted.

        "accounts": ["solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp:7S3P4HxJpyyigGzodYwHtCxZyUQe9JiBMHyRWXArAaKv"]
      }
    },      
  }
}
```

# Proposal

## Language
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in uppercase in this document are to be interpreted as described in RFC 2119.

## Definitions

**Multichain API**: a new [Wallet API](https://docs.metamask.io/wallet/concepts/wallet-api/) that is generalized for interactions between dapps and wallets in a multichain environment

**network**: Decentralized networks where control, validation, and decision-making processes are distributed across multiple nodes rather than centralized in a single entity. The term will refer to a specific decentralized network that has a single unique identifier. Networks may be implemented as a blockchain or another distributed ledger technology. 

**node**: An RPC node that facilitates interactions with a decentralized network.

> **Wallet API**: [The JSON-RPC interface](https://docs.metamask.io/wallet/concepts/apis/) that dapps can use to programmatically interact with MetaMask wallet clients.

Reviewer 3: Do you want to differentiate between the Wallet API's provider API and JSON-RPC API? If so, is the Multichain API more equivalent to the provider API?

Author: I'm not sure the Multichain API precisely maps to the current JSON-RPC API, but many concepts around the provider API (such as EIP-1193 patterns) are going away with the introduction of the Multichain API. The new API is less reliant on a JavaScript object interface and primarily involves JSON-RPC request/response.

> **Note:** a `sessionId` parameter is not supported and will be ignored, if included.

### CAIP-311 - Notification of Changes to Authorization Scopes
A dapp can listen for [CAIP-311](https://chainagnostic.org/CAIPs/caip-311) `wallet_sessionChanged` events in order to get notified when there are changes to its authorization scopes.

> **Note:** A `sessionId` will not be included as part of the event.

### CAIP-285 - Revoke Authorization Scopes
A dapp can make a [CAIP-285](https://chainagnostic.org/CAIPs/caip-285) `wallet_revokeSession` request to revoke all of its authorization scopes.

> **Note:** a `sessionId` parameter is not supported and will be ignored, if included.

### API Specification Document
An OpenRPC [specification for the Multichain API](https://github.com/MetaMask/api-specs/blob/main/multichain/openrpc.yaml) can be found in the [api-specs](https://github.com/MetaMask/api-specs) repository.

## Caveats
The CAIPs referenced in this proposal are still in Draft or Review status with CASA and may be subject to material changes.

### API Access

> As part of the Multichain API, MetaMask is leading the adoption of alternative methods for browser extension wallet communication. The current practice for browser extension wallets is to inject a javascript object in order to communicate with each site. But this approach comes with privacy, security, reliability, and performance drawbacks.

Reviewer 3: ```suggestion
As part of the Multichain API, MetaMask is leading the adoption of alternative methods for browser extension wallet communication. Currently, browser extension wallets inject a JavaScript object to communicate with each site. But this approach comes with privacy, security, reliability, and performance drawbacks.
```

## Usage Example

### Request
An example structure for a JSON-RPC request that a dapp would send to request authorization(s) from MetaMask:

```js
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "wallet_createSession",
  "params": {
    "optionalScopes": {
      "wallet:eip155": {
        "methods": ["wallet_addEthereumChain"],
        "notifications": [],
      },
      "eip155:1": {
        "methods": ["eth_sendTransaction","eth_call","eth_getBalance","eth_blockNumber","eth_getTransactionCount","wallet_watchAsset","eth_subscribe"],
        "notifications": ["eth_subscription"],
      },
      "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp": {
        "methods": ["signAndSendTransaction","signAllTransactions","signMessage"],
        "notifications": [],
      }
      "bip122:000000000019d6689c085ae165831e93": {
        ...
      }
    },
    "scopedProperties": {
      // Scope-keyed proposals or declarations
      "eip155:1": {
        "extension_foo": "bar"    
      }
    },
  },
}
```

### Response
An example structure for the corresponding JSON-RPC response that a dapp would receive from the wallet:

```js
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "sessionScopes": { 

>       "eip155:wallet": {

Reviewer: ```suggestion
      "wallet:eip155": {
```

          "methods": ["wallet_addEthereumChain"],
          "notifications": [],

>           "accounts": ["eip155:wallet:0x0910e12C68d02B561a34569E1367c9AAb42bd810"]

Reviewer: ```suggestion
          "accounts": ["wallet:eip155:0x0910e12C68d02B561a34569E1367c9AAb42bd810"]
```

- The MetaMask Browser Extension will adopt an alternative method of communication through the [externally_connectable](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable) feature, which is now supported by most modern web browsers with extension frameworks.
- When a site is accessed through a browser that does not support [externally_connectable](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable), the Multichain API will be available through an alternative communication method. 

In summary, the following methods will solely be accessible through the new communication methods:

- `wallet_createSession` (CAIP-25)
- `wallet_sessionChanged` (CAIP-311)
- `wallet_getSession` (CAIP-312)
- `wallet_revokeSession` (CAIP-285)
- `wallet_invokeMethod` (CAIP-27)
- `wallet_notify` (CAIP-319)

> (API Delivery & Wallet Discovery CAIPs to be added in this section)

Reviewer: Assuming this will be dealt with later

Author: Yes. I think we should wait to link once we at least have draft CAIPs posted.
